README.md
## Lazy and infinite linked lists for JavaScript

[![license](https://img.shields.io/badge/license-ISC-blue.svg)](https://github.com/sjsyrek/lazy-linked-lists/blob/master/LICENSE.txt)
[![build status](https://travis-ci.org/sjsyrek/lazy-linked-lists.svg?branch=master)](https://travis-ci.org/sjsyrek/lazy-linked-lists)
[![test coverage](https://codeclimate.com/github/sjsyrek/lazy-linked-lists/badges/coverage.svg)](https://codeclimate.com/github/sjsyrek/lazy-linked-lists/coverage)
[![bitHound score](https://www.bithound.io/github/sjsyrek/lazy-linked-lists/badges/score.svg)](https://www.bithound.io/github/sjsyrek/lazy-linked-lists)
[![dependencies status](https://david-dm.org/sjsyrek/lazy-linked-lists.svg)](https://david-dm.org/sjsyrek/lazy-linked-lists)
[![devDependencies status](https://david-dm.org/sjsyrek/lazy-linked-lists/dev-status.svg)](https://david-dm.org/sjsyrek/lazy-linked-lists?type=dev)
[![dependency status](https://dependencyci.com/github/sjsyrek/lazy-linked-lists/badge)](https://dependencyci.com/github/sjsyrek/lazy-linked-lists)
[![style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/334/badge)](https://bestpractices.coreinfrastructure.org/projects/334)

[![NPM](https://nodei.co/npm/lazy-linked-lists.png?downloads=true)](https://nodei.co/npm/lazy-linked-lists/)

> A spectre is haunting ECMAScript—the spectre of tail call optimization.
> — Karl Marxdown

## About

This library is adapted from [maryamyriameliamurphies.js](https://github.com/sjsyrek/maryamyriameliamurphies.js) with several modifications. It includes functions for creating both eagerly and lazily-evaluated linked list data structures, including infinite lists, and a core subset of functions for working with them. Unlike **maryamyriameliamurphies.js**, however, **lazy-linked-lists** does not implement function currying, partial application, or type checking. It does, however, implement most of the standard [Haskell](https://www.haskell.org) type classes as instance methods, and it also implements the [ES2015 iteration protocols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols), so you can use them in `for...of` loops or otherwise as regular iterators.

The lazy lists provided by this library are implemented using the new `Proxy` API in ES2015. Briefly, the lists returned from most of the list constructors are actually hidden behind proxy objects that trap references to their "tail" property, so that list elements, produced by an ES2015 generator function, are only evaluated on demand. Obviously, if you request the entire tail (by, for example, calling the `length()` function on a list), then the entire tail will be evaluated. You will want to avoid doing that with infinite lists.

Note that as of this writing, most implementations of the ES2015 standard do not yet support [proper tail calls](http://www.2ality.com/2015/06/tail-call-optimization.html). But support is on its way! The newest versions of node and Safari have already rolled it out, and other vendors are surely not far behind. See the top line of this [compatibility chart](https://kangax.github.io/compat-table/es6/) to track the progress of the feature that will make recursively defined, fully-functional, high octane linked lists in JavaScript a reality. Until that fateful day, however, you may be limited to lists of only 10,000 elements or so.

For further details, see the [documentation](http://sjsyrek.github.io/maryamyriameliamurphies.js/) for **maryamyriameliamurphies.js**.

##### [Try it now with Tonic](https://tonicdev.com/npm/lazy-linked-lists)

## Examples

Linked list, eagerly evaluated:
```js
const lst = list(1,2,3,4,5,6,7,8,9,10);
lst.valueOf(); // => '[1:2:3:4:5:6:7:8:9:10:[]]'
```

Linked list, lazily evaluated:
```js
const lst = listRange(1, 10);
lst.valueOf(); // => '[1:2:3:4:5:6:7:8:9:10:[]]'
```

Linked list, lazily evaluated with a user-defined step function:
```js
const lst1 = listRangeBy(0, 100, x => x + 10);
const lst2 = listRangeBy(10, 0, x => x - 1);
lst1.valueOf(); // => '[0:10:20:30:40:50:60:70:80:90:100:[]]'
lst2.valueOf(); // => '[10:9:8:7:6:5:4:3:2:1:[]]'
```

Iterating over a linked list:
```js
const lst = listRange(1, 5);
for (let value of lst) { console.log(value); }
// 1
// 2
// 3
// 4
// 5
```

Infinite list:
```js
const lst = listInf(1);
take(10, lst).valueOf(); // => '[1:2:3:4:5:6:7:8:9:10:[]]'
lst.valueOf(); // => RangeError: Maximum call stack size exceeded
```

Infinite list with a user-defined step function:
```js
const lst = listInfBy(0, x => x + 10);
take(11, lst).valueOf(); // => '[0:10:20:30:40:50:60:70:80:90:100:[]]'
```

Haskell-style type class action:
```js
const lst1 = list(1,2,3);
const lst2 = list(3,2,1);

// Eq
lst1.isEq(list(1,2,3)); // => true
lst1.isEq(lst2); // => false

// Ord
lst1.compare(lst2); // => Symbol()
lst1.compare(lst2) === LT; // => true
lst1.isLessThan(lst2); // => true
lst1.isGreaterThan(lst2); // => false

// Monoid
lst1.mappend(lst1.mempty()); // => '[1:2:3:[]]'
lst1.mappend(lst2); // => '[1:2:3:4:5:6:[]]'

// Foldable
lst1.foldr((x,y) => x * y, 1); // => 6

// Traversable
lst1.traverse(x => list(x * 10)); // => '[[10:20:30:[]]:[]]'

// Functor
lst1.fmap(x => x * 10); // => '[10:20:30:[]]'

// Applicative
const f = x => x * 10;
const fs1 = list(f);
const fs2 = list(f,f,f);
fs1.ap(lst1); // => '[10:20:30:[]]'
fs2.ap(lst1); // => '[10:20:30:10:20:30:10:20:30:[]]'

// Monad
lst1.flatMap(x => list(x * 10)); // => '[10:20:30:[]]'
lst1.then(lst2); // => '[4:5:6:4:5:6:4:5:6:[]]'
const stringify = x => list(`${x}`);
lst1.flatMap(x => list(x, x * 10, x * x)).flatMap(stringify); // => '[110122043309]'
```

Other fun stuff:
```js
const lst1 = listInfBy(0, x => x + 2);
const lst2 = take(10, lst1);
const lst3 = reverse(lst2);
const lst4 = sort(lst3);
lst3.valueOf(); // => '[18:16:14:12:10:8:6:4:2:0:[]]'
lst4.valueOf(); // => '[0:2:4:6:8:10:12:14:16:18:[]]'

const lst5 = iterate(x => x * 2, 1);
const lst6 = take(10, lst5);
lst6.valueOf(); // => '[1:2:4:8:16:32:64:128:256:512:[]]'
index(lst6, 10); // => Error: *** Exception: index: range error
index(lst5, 10); // => 1024

const lst7 = repeat(3);
const lst8 = take(10, lst7);
lst8.valueOf(); // => '[3:3:3:3:3:3:3:3:3:3:[]]'
index(lst7, 100); // => 3

const lst9 = replicate(10, 3);
lst9.valueOf(); // => [3:3:3:3:3:3:3:3:3:3:[]]

const lst10 = list(1,2,3);
const lst11 = cycle(lst10);
const lst12 = take(9, lst11);
lst12.valueOf(); // => [1:2:3:1:2:3:1:2:3:[]]
index(lst11, 99); // => 1
index(lst11, 100); // => 2
index(lst11, 101); // => 3
```

See also the files in the example directory.

### How to install and use

- [Install with npm](https://www.npmjs.com/package/lazy-linked-lists) `npm install --save-dev lazy-linked-lists`. _Do not_ install this package globally.
- If you're transpiling >=ES2015 code with [Babel](http://babeljs.io), put `import * as lazy from 'lazy-linked-lists';` at the top of your script files.
- Or, to pollute your namespace, import functions individually: `import {listRange, listRangeBy} from 'lazy-linked-lists';`.
- Or, if you aren't transpiling (or you're old school), use node's require syntax: `const lazy = require('lazy-linked-lists');`.

### How to develop

- [Fork this repo](https://help.github.com/articles/fork-a-repo/) and [clone it locally](https://help.github.com/articles/cloning-a-repository/).
- `npm install` to download the dependencies.
- `npm run compile` to run Babel on ES2015 code in `./source` and output transpiled ES5 code to `./distribution`.
- `npm run lint` to run ESlint to check the source code for errors.
- `npm test` to run Mocha on the test code in `./test`.
- `npm run cover` to run nyc on the source code and generate testing coverage reports.
- `npm run clean` to delete all files in `./distribution`.

#### See also

* [lazy.js](https://github.com/dtao/lazy.js)
* [node-lazy](https://github.com/pkrumins/node-lazy)
* [lazy-list](https://github.com/luochen1990/lazy-list)
* [js-list-lazy](https://github.com/dankogai/js-list-lazy)
* [fantasy-land](https://github.com/fantasyland/fantasy-land)
* [js-algebraic](https://github.com/tel/js-algebraic)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.2"></a>
### 0.0.1 (2016-08-20)

#### Features

* create lazy-linked-lists package ([fde2764d10d30ca299aafd956f02a2d29ba8ba74](https://github.com/sjsyrek/lazy-linked-lists/commit/fde2764d10d30ca299aafd956f02a2d29ba8ba74))


test/index.js
/**
 * lazy-linked-lists
 * Lazy and infinite linked lists for JavaScript.
 *
 * test/index.js
 *
 * Top level index for unit tests.
 * @license ISC
 */

import 'should';

import './units';


example/sqrt.js
/**
 * lazy-linked-lists
 * Lazy and infinite linked lists for JavaScript.
 *
 * example/sqrt.js
 *
 * Examples of usage for lazy-linked-lists.
 * @license ISC
 */

import * as lazy from '../source'

/**
 * Compute the square root of a number using infinite lists with the Newton-Raphson method.
 * Adapted from "Why Functional Programming Matters" by John Hughes.
 * @param {number} a0 - initial estimate
 * @param {number} eps - tolerance
 * @param {number} n - number
 * @returns {number} The square root of `n`
 * @kind function
 * @example
 * sqrt(1,0,2);           // => 1.414213562373095
 * sqrt(1,0,144);         // => 12
 * relativeSqrt(1,0,144); // optimized version for very small and very large numbers
 */
export const sqrt = (a0, eps, n) => within(eps, lazy.iterate(next.bind(null, n), a0))

export const relativeSqrt = (a0, eps, n) => relative(eps, lazy.iterate(next.bind(null, n), a0))

const next = (n, x) => (x + n / x) / 2

const within = (eps, rest) => {
  let a = lazy.index(rest, 0)
  let b = lazy.index(rest, 1)
  return Math.abs(a - b) <= eps ? b : within(eps, lazy.drop(1, rest))
}

const relative = (eps, rest) => {
  let a = lazy.index(rest, 0)
  let b = lazy.index(rest, 1)
  return Math.abs(a - b) <= eps * Math.abs(b) ? b : relative(eps, lazy.drop(1, rest))
}


example/example.js
/**
 * lazy-linked-lists
 * Lazy and infinite linked lists for JavaScript.
 *
 * example.js
 *
 * Examples of usage for lazy-linked-lists.
 * @license ISC
 */

import * as lazy from '../source';

/**
 * Compute the square root of a number using the Newton-Raphson method.
 * Adapted from "Why Functional Programming Matters" by John Hughes.
 * @param {number} a0 - initial estimate
 * @param {number} eps - tolerance
 * @param {number} n - number
 * @returns {number} The square root of `n`
 * @kind function
 * @example
 * sqrt(1,0,2);           // => 1.414213562373095
 * sqrt(1,0,144);         // => 12
 * relativeSqrt(1,0,144); // optimized version for very small and very large numbers
 */
const sqrt = (a01, eps0, n) => within(eps, lazy.iterate(next.bind(null, n), a0));

const relativeSqrt = (a0, eps, n) => relative(eps, lazy.iterate(next.bind(null, n), a0));

const next = (n, x) => (x + n/x) / 2;

const within = (eps, rest) => {
  let a = lazy.index(rest, 0), b = lazy.index(rest, 1);
  return Math.abs(a - b) <= eps ? b : within(eps, lazy.drop(1, rest));
}

const relative = (eps, rest) => {
  let a = lazy.index(rest, 0), b = lazy.index(rest, 1);
  return Math.abs(a - b) <= eps * Math.abs(b) ? b : relative(eps, lazy.drop(1, rest));
}


source/lib.js
  /**
 * lazy-linked-lists
 * Lazy and infinite linked lists for JavaScript.
 *
 * source/lib.js
 *
 * Core functions for lazy-linked-lists.
 * @license ISC
 */

import {List} from './list'

import {
  LT,
  GT,
  EQ
} from './ord'

import {
  EmptyListError,
  OutOfRangeError
} from './error'

export const emptyList = new List()

/**
 * Create a new `List` from a series of zero or more values.
 * @param {...*} as - Values to put into a new `List`
 * @returns {List} The new `List`
 * @kind function
 * @example
 * list(1,2,3); // => [1:2:3:[]]
 */
export const list = (...as) => as.length === 0 ? emptyList : new List(as.shift(), list(...as))

/**
 * Build a `List` from a range of values using lazy evaluation (i.e. each successive value is only
 * computed on demand, making infinite lists feasible). To supply your own function for determining
 * the increment, use `listRangeBy`.
 * @param {*} start - The starting value
 * @param {*} end - The end value
 * @returns {List} A `List` that will be evaluated lazily
 * @kind function
 */
export const listRange = (start, end) => {
  if (start === end) { return list(start) }
  if (start > end) { return listRangeBy(start, end, x => x - 1) }
  return listRangeBy(start, end, x => x + 1)
}

/**
 * Build a `List` from a range of values using lazy evaluation and incrementing it using a given
 * step function.
 * @param {*} start - The starting value
 * @param {*} end - The end value
 * @param {Function} step - The increment function
 * @returns {List} A `List` that will be evaluated lazily
 * @kind function
 */
export const listRangeBy = (start, end, step) => {
  if (start === end) { return list(start) }
  let x = start
  const xs = list(x)
  const listGenerator = function* () {
    x = step(x)
    while (start < end ? x < end : x > end) {
      yield list(x)
      x = step(x)
    }
  }
  const gen = listGenerator()
  const handler = {
    get: function (target, prop) {
      if (prop === `tail` && isEmpty(tail(target))) {
        const next = gen.next()
        if (next.done === false) { target[prop] = () => new Proxy(next.value, handler) }
      }
      return target[prop]
    }
  }
  const proxy = new Proxy(xs, handler)
  return proxy
}

/**
 * Append one `List` to another.
 * @param {List} xs - A `List`
 * @param {List} ys - A `List`
 * @returns {List} A new `List`, the result of appending `xs` to `ys`
 * @kind function
 * @example
 * const lst1 = list(1,2,3);
 * const lst2 = list(4,5,6);
 * listAppend(lst1, lst2);   // => [1:2:3:4:5:6:[]]
 */
export const listAppend = (xs, ys) => {
  if (isEmpty(xs)) { return ys }
  if (isEmpty(ys)) { return xs }
  return cons(head(xs), listAppend(tail(xs), ys))
}

/**
 * Create a new `List` from a head and tail.
 * @param {*} x - Any value, the head of the new list
 * @param {List} xs - A `List`, the tail of the new list
 * @returns {List} The new `List`, constructed from `x` and `xs`
 * @kind function
 * @example
 * const lst = list(4,5,6);
 * cons(3)(lst);            // => [3:4:5:6:[]]
 */
export const cons = (x, xs) => new List(x, xs)

/**
 * Extract the first element of a `List`.
 * @param {List} xs - A `List`
 * @returns {*} The head of the `List`
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * head(lst);               // => 1
 */
export const head = xs => {
  if (isEmpty(xs)) { throw new EmptyListError(head) }
  return xs.head()
}

/**
 * Extract the last element of a `List`.
 * @param {List} xs - A `List`
 * @returns {*} The last element of the `List`
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * last(lst);               // => 3
 */
export const last = xs => {
  if (isEmpty(xs)) { throw new EmptyListError(last) }
  return isEmpty(tail(xs)) ? head(xs) : last(tail(xs))
}

/**
 * Extract the elements after the head of a `List`.
 * @param {List} xs - A `List`
 * @returns {List} The tail of the `List`
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * tail(lst);               // => [2:3:[]]
 */
export const tail = xs => {
  if (isEmpty(xs)) { throw new EmptyListError(tail) }
  return xs.tail()
}

/**
 * Return all the elements of a `List` except the last one.
 * @param {List} xs - A `List`
 * @returns {List} A new `List`, without the original list's last element
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * init(lst);               // => [1:2:[]]
 */
export const init = xs => {
  if (isEmpty(xs)) { throw new EmptyListError(init) }
  return isEmpty(tail(xs)) ? emptyList : cons(head(xs), init(tail(xs)))
}

/**
 * Return the length of a `List`.
 * @param {List} xs - A `List`
 * @returns {number} The length of the `List`
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * length(lst);             // => 3
 */
export const length = xs => {
  const lenAcc = (xs, n) => isEmpty(xs) ? n : lenAcc(tail(xs), n + 1)
  return lenAcc(xs, 0)
}

/**
 * Determine whether a given object is a `List`.
 * @param {*} a - Any object
 * @returns {boolean} `true` if the object is a `List` and `false` otherwise
 * @kind function
 */
export const isList = a => a instanceof List

/**
 * Check whether a `List` is empty. Returns `true` if the `List` is empty or false if it is
 * non-empty.
 * @param {*} xs - A `List`
 * @returns {boolean} `true` if the `List` is empty, `false` if it is non-empty
 * @kind function
 * @example
 * isEmpty([]);              // => true
 * isEmpty([[]]);            // => false
 * isEmpty(emptyList);       // => true
 * isEmpty(list(emptyList)); // => false
 */
export const isEmpty = xs => xs === emptyList

/**
 * Convert an array into a `List`.
 * @param {Array.<*>} arr - An array to convert into a `List`
 * @returns {List} A new `List`, the converted array
 * @kind function
 * @example
 * const arr = [1,2,3];
 * fromArrayToList(arr); // => [1:2:3:[]]
 */
export const fromArrayToList = arr => list(...arr)

/**
 * Convert a `List` into an array.
 * @param {List} xs - A `List` to convert into an array
 * @returns {Array} A new array, the converted `List`
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * fromListToArray(lst);    // => [1,2,3]
 */
export const fromListToArray = xs => isEmpty(xs) ? [] : [head(xs)].concat(fromListToArray(tail(xs)))
/**
 * Convert a `List` into a string.
 * @param {List} xs - A `List` to convert into a string
 * @returns {string} A new string, the converted `List`
 * @kind function
 * @example
 * const str = list('a','b','c');
 * fromListToString(str);         // => "abc"
 */
export const fromListToString = xs => fromListToArray(xs).join(``)

/**
 * Convert a string into a `List`.
 * @param {string} str - A string to convert into a `List`
 * @returns {List} A new `List`, the converted string
 * @kind function
 * @example
 * const str = `abc`;
 * fromStringToList(str); // => [abc]
 */
export const fromStringToList = str => fromArrayToList(str.split(``))

/**
 * Concatenate the elements in a `List` of lists.
 * @param {List} xss - A `List` of lists
 * @returns {List} The concatenated `List`
 * @kind function
 * @example
 * const lst1 = list(1,2,3);
 * const lst2 = list(4,5,6);
 * const lst3 = list(7,8,9);
 * const xss = list(lst1, lst2, lst3); // [[1:2:3:[]]:[4:5:6:[]]:[7:8:9:[]]:[]]
 * concat(xss);                        // => [1:2:3:4:5:6:7:8:9:[]]
 */
export const concat = xss => {
  if (isEmpty(xss)) { return emptyList }
  const x = head(xss)
  const xs = tail(xss)
  return listAppend(x, concat(xs))
}

/**
 * Return the value from a `List` at the specified index, starting at 0.
 * @param {List} as - The `List` to index into
 * @param {number} n - The index to return
 * @returns {*} The value at the specified index
 * @kind function
 * @example
 * const lst = list(1,2,3,4,5);
 * index(lst, 3));              // => 4
 */
export const index = (as, n) => {
  if (n < 0 || isEmpty(as)) { throw new OutOfRangeError(head) }
  const x = head(as)
  const xs = tail(as)
  if (n === 0) { return x }
  return index(xs, n - 1)
}

/**
 * Return the `List` of elements in a `List` for which a function `f` returns `true`.
 * @param {Function} f - The filter function. Must return a `boolean`
 * @param {List} as - The `List` to filter
 * @returns {List} The filtered `List`
 * @kind function
 * @example
 * const lst = listRange(1,50);
 * const f = x => and(odd(x), greaterThan(x, 10));
 * filter(f, lst); // => [11:13:15:17:19:21:23:25:27:29:31:33:35:37:39:41:43:45:47:49:[]]
 */
export const filter = (f, as) => {
  if (isEmpty(as)) { return emptyList }
  const x = head(as)
  const xs = tail(as)
  if (f(x) === true) { return cons(x, filter(f, xs)) }
  return filter(f, xs)
}

/**
 * Map a function over a `List` and put the results into a new `List`.
 * @param {Function} f - The function to map
 * @param {List} as - The `List` to map over
 * @returns {List} A `List` of results
 * @kind function
 * @example
 * const lst = list(1,2,3,4,5);
 * const f = x => x * 3;
 * map(f, lst));                 // => [3:6:9:12:15:[]]
 */
export const map = (f, as) => {
  if (isEmpty(as)) { return emptyList }
  const x = f(head(as))
  const xs = tail(as)
  return cons(x, map(f, xs))
}

/**
 * Reverse the elements of a `List` and return them in a new list.
 * @param {List} xs - A `List`
 * @returns {List} The reversed `List`
 * @kind function
 * @example
 * const lst = list(1,2,3,4,5);
 * reverse(lst);                // => [5:4:3:2:1:[]]
 */
export const reverse = xs => {
  const r = (xs, a) => isEmpty(xs) ? a : r(tail(xs), cons(head(xs), a))
  return r(xs, emptyList)
}

/**
 * Sort a list using regular value comparison. Use `sortBy` to supply your own comparison
 * function. Uses a merge sort algorithm.
 * @param {List} xs - The `List` to sort
 * @returns {List} - The sorted `List` (the original list is unmodified)
 * @kind function
 * @example
 * const lst1 = list(20,19,18,17,16,15,14,13,12,11,10,1,2,3,4,5,6,7,8,9);
 * sort(lst1); // => [1:2:3:4:5:6:7:8:9:10:11:12:13:14:15:16:17:18:19:20:[]]
 * const f = x => x + 1;
 * const lst2 = reverse(listRange(1, 11, f)); // [10:9:8:7:6:5:4:3:2:1:[]]
 * sort(lst2);                           // => [1:2:3:4:5:6:7:8:9:10:[]]
 */
export const sort = xs => sortBy((a, b) => a === b ? EQ : (a < b ? LT : GT), xs)

/**
 * Sort a list using a comparison function of your choice. Uses a merge sort algorithm.
 * @param {Function} cmp - The comparison function—must return an `Ordering`
 * @param {List} as - The `List` to sort
 * @returns {List} The sorted `List` (the original list is unmodified)
 * @kind function
 */
export const sortBy = (cmp, as) => {
  const sequences = as => {
    if (isEmpty(as)) { return list(as) }
    let xs = tail(as)
    if (isEmpty(xs)) { return list(as) }
    const a = head(as)
    const b = head(xs)
    xs = tail(xs)
    if (cmp(a, b) === GT) { return descending(b, list(a), xs) }
    return ascending(b, cons.bind(null, a), xs)
  }
  const descending = (a, as, bbs) => {
    if (isEmpty(bbs)) { return cons(cons(a, as), sequences(bbs)) }
    const b = head(bbs)
    const bs = tail(bbs)
    if (cmp(a, b) === GT) { return descending(b, cons(a, as), bs) }
    return cons(cons(a, as), sequences(bbs))
  }
  const ascending = (a, as, bbs) => {
    if (isEmpty(bbs)) { return cons(as(list(a)), sequences(bbs)) }
    const b = head(bbs)
    const bs = tail(bbs)
    const ys = ys => as(cons(a, ys))
    if (cmp(a, b) !== GT) { return ascending(b, ys, bs) }
    return cons(as(list(a)), sequences(bbs))
  }
  const mergeAll = xs => {
    if (isEmpty(tail(xs))) { return head(xs) }
    return mergeAll(mergePairs(xs))
  }
  const mergePairs = as => {
    if (isEmpty(as)) { return as }
    let xs = tail(as)
    if (isEmpty(xs)) { return as }
    const a = head(as)
    const b = head(xs)
    xs = tail(xs)
    return cons(merge(a, b), mergePairs(xs))
  }
  const merge = (as, bs) => {
    if (isEmpty(as)) { return bs }
    if (isEmpty(bs)) { return as }
    const a = head(as)
    const as1 = tail(as)
    const b = head(bs)
    const bs1 = tail(bs)
    if (cmp(a, b) === GT) { return cons(b, merge(as, bs1)) }
    return cons(a, merge(as1, bs))
  }
  return mergeAll(sequences(as))
}

/**
 * Return the prefix of a `List` of a given length.
 * @param {number} n - The length of the prefix to take
 * @param {List} as - The `List` to take from
 * @returns {List} A new `List`, the desired prefix of the original list
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * take(2, lst);            // => [1:2:[]]
 */
export const take = (n, as) => {
  if (n <= 0) { return emptyList }
  if (isEmpty(as)) { return emptyList }
  const x = head(as)
  const xs = tail(as)
  return cons(x, take(n - 1, xs))
}

/**
 * Return the suffix of a `List` after discarding a specified number of values.
 * @param {number} n - The number of values to drop
 * @param {List} as - The `List` to drop values from
 * @returns {List} A new `List`, the desired suffix of the original list
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * drop(2, lst);            // => [3:[]]
 */
export const drop = (n, as) => {
  if (n <= 0) { return as }
  if (isEmpty(as)) { return emptyList }
  const xs = tail(as)
  return drop(n - 1, xs)
}

/**
 * Return the longest prefix (possibly empty) of a `List` of values that satisfy a predicate
 * function.
 * @param {Function} p - The predicate function (should return `boolean`)
 * @param {List} as - The `List` to take from
 * @returns {List} The `List` of values that satisfy the predicate function
 * @kind function
 * @example
 * const lst = list(1,2,3,4,1,2,3,4);
 * const f = x => x < 3;
 * takeWhile(f, lst);                 // => [1:2:[]]
 */
export const takeWhile = (p, as) => {
  if (isEmpty(as)) { return emptyList }
  const x = head(as)
  const xs = tail(as)
  const test = p(x)
  if (test === true) { return cons(x, takeWhile(p, xs)) }
  return emptyList
}

/**
 * Drop values from a `List` while a given predicate function returns `true` for each value.
 * @param {Function} p - The predicate function (should return `boolean`)
 * @param {List} as - The `List` to drop values from
 * @returns {List} The `List` of values that do not satisfy the predicate function
 * @kind function
 * @example
 * const lst = list(1,2,3,4,5,1,2,3);
 * const f = x => x < 3;
 * dropWhile(f, lst);                 // => [3:4:5:1:2:3:[]]
 */
export const dropWhile = (p, as) => {
  if (isEmpty(as)) { return emptyList }
  const x = head(as)
  const xs = tail(as)
  const test = p(x)
  if (test === true) { return dropWhile(p, xs) }
  return as
}

/**
 * Generate an infinite `List`. Use `listInfBy` to supply your own step function.
 * @param {*} start - The value with which to start the `List`
 * @returns {List} An infinite `List` of consecutive values, incremented from `start`
 * @kind function
 */
export const listInf = start => listInfBy(start, x => x + 1)

/**
 * Generate an infinite `List`, incremented using a given step function.
 * @param {*} start - The value with which to start the `List`
 * @param {Function} step - A unary step function
 * @returns {List} An infinite `List` of consecutive values, incremented from `start`
 * @kind function
 */
export const listInfBy = (start, step) => listRangeBy(start, Infinity, step)

/**
 * Return an infinite `List` of repeated applications of a function to a value.
 * @param {Function} f - The function to apply
 * @param {*} x - The value to apply the function to
 * @returns {List} An infinite `List` of repeated applications of `f` to `x`
 * @kind function
 * @example
 * const f = x => x * 2;
 * const lst = iterate(f, 1);
 * take(10, lst);             // => [1:2:4:8:16:32:64:128:256:512:[]]
 * index(lst, 10);            // => 1024
 */
export const iterate = (f, x) => listInfBy(x, x => f(x))

/**
 * Build an infinite `List` of identical values.
 * @param {*} a - The value to repeat
 * @returns {List} The infinite `List` of repeated values
 * @kind function
 * @example
 * const lst = repeat(3);
 * take(10, lst);         // => [3:3:3:3:3:3:3:3:3:3:[]]
 * index(lst, 100);       // => 3
 */
export const repeat = a => cons(a, listInfBy(a, a => a))

/**
 * Return a `List` of a specified length in which every value is the same.
 * @param {number} n - The length of the `List`
 * @param {*} x - The value to replicate
 * @returns {List} The `List` of values
 * @kind function
 * @example
 * replicate(10, 3); // => [3:3:3:3:3:3:3:3:3:3:[]]
 */
export const replicate = (n, x) => take(n, repeat(x))

/**
 * Return the infinite repetition of a `List` (i.e. the "identity" of infinite lists).
 * @param {List} as - A finite `List`
 * @returns {List} A circular `List`, the original `List` infinitely repeated
 * @kind function
 * @example
 * const lst = list(1,2,3);
 * const c = cycle(lst);
 * take(9, c);              // => [1:2:3:1:2:3:1:2:3:[]]
 * index(c, 100);           // => 2
 */
export const cycle = as => {
  if (isEmpty(as)) { throw new EmptyListError(cycle) }
  let x = head(as)
  let xs = tail(as)
  const c = list(x)
  /* eslint no-constant-condition: ["error", { "checkLoops": false }] */
  const listGenerator = function* () {
    do {
      x = isEmpty(xs) ? head(as) : head(xs)
      xs = isEmpty(xs) ? tail(as) : tail(xs)
      yield list(x)
    } while (true)
  }
  const gen = listGenerator()
  const handler = {
    get: function (target, prop) {
      if (prop === `tail` && isEmpty(tail(target))) {
        const next = gen.next()
        target[prop] = () => new Proxy(next.value, handler)
      }
      return target[prop]
    }
  }
  const proxy = new Proxy(c, handler)
  return proxy
}
